### ISSUE TITLE
[BUG] Backing Image Data Inconsistency if it's Exported from a Backing Image Backed Volume

### ISSUE THREAD

**Issue by WebberHuang1118** [harvester/harvester#4900]
**Describe the bug**
The volume encounters a data integrity issue, if it's from a backing image, and such backing image is exported by a volume backed by another backing image.

**To Reproduce**
1. Preparing a Harvester v1.2.0-head cluster (for using volume as an OS root disk)
2. Uploading Ubuntu jammy cloud image ([latest version](https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img)), **img1**
3. Creating volume **vol1** from **img1**
4. Creating a VM with **vol1** as the root disk, and writing some data
5. Issuing sync on filesystem, e.q. `$ sudo sync`
6. Stopping VM, and waiting until **vol1** is completely detached
7. Exporting **vol1** as **img2**
8. Creating  **vol2** from **img2**
9. Comparing data content between **vol1** and **vol2**, the data is inconsistent
10. Running fsck on **vol2**, fsck will report filesystem errors and require data repair

**Expected behavior**
The data of **vol1** and **vol2** should be consistent.

**Support bundle for troubleshooting**
[supportbundle_066aa8ed-a12b-4ccc-971c-59404ee44cd4_2023-10-16T07-46-44Z.zip](https://github.com/longhorn/longhorn/files/12913647/supportbundle_066aa8ed-a12b-4ccc-971c-59404ee44cd4_2023-10-16T07-46-44Z.zip)

**Environment**
 - Longhorn version: v1.4.3
 - Harvester version: v1.2.0-head
 - Number of Longhorn volumes in the cluster: 2
 - Impacted Longhorn resources:
   - Volume names: pvc-b8bd9013-e8e9-449d-9c25-6c6c0824d0b3, pvc-de64dd95-b0c1-4ce9-864a-88d38e06d9d3

** Additional context**
- If **vol1** is not backed by backing image, there is no data inconsistency found between **vol1** and **vol2**
- A single node Harvester cluster should be enough to reproduce this issue

**Comment by WebberHuang1118** [comment]
Should be tacked by LH issue https://github.com/longhorn/longhorn/issues/6899

**Comment by WebberHuang1118** [comment]
I found a method to reproduce a similar case without Harvester-specific operations (no VM operation, only creating PVC from backing image and I/O access)

Detailed steps:
- Uploading Ubuntu jammy cloud image ([latest version](https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img)), **img1**
- Creating PVC **os-vol** from **img1** with claim size 10GB and 1 replica number
- Applying the following yaml to mount **os-vol** as block mode in the pod
  ```
  apiVersion: v1
  kind: Pod
  metadata:
    name: block-volume-test
    namespace: default
  spec:
    containers:
    - name: volume-test
      image: ubuntu
      imagePullPolicy: IfNotPresent
      securityContext:
        privileged: true
      command: ["/bin/sleep"]
      args: ["3600"]
      volumeDevices:
      - devicePath: /dev/os-vol
        name: os-vol
    volumes:
      - name: os-vol
        persistentVolumeClaim:
          claimName: os-vol
  ```  
- Issuing filesystem expand and I/O on **os-vol**
  -  Enter pod **block-volume-test**
  -  Install fdisk
     - $ apt update; apt install -y fdisk
  - Expanding filesystem on **os-vol**
    - Check device (/dev/sda for exmaple) mapped to **os-vol** by device number
    - $ fdisk /dev/sda
      - d 1 n 1 default default No w
    - $ resize2fs -fp /dev/sda1
  - Mounting /dev/sda1 and writing I/O into it
    - $ mkdir /data; $ mount -t ext4 /dev/sda1 /data
    - $ dd if=/dev/zero of=/data/file bs=1M count=2048
  - Flushing data to storage for **os-vol**
    - $ umount /data
  -  Delete pod **block-volume-test**, and waiting until **os-vol**'s LH volume detach
- Exporting **os-vol** as image **os-vol.img**
- Creating PVC **copy-vol** from **os-vol.img** with claim size 10GB
- Comparing data content between **os-vol** and **copy-vol** with `$cmp -l os-vol copy-vol`
  - The data content is different between the two PVCs
- Finding filesystem error on **copy-vol** with` $e2fsck -c /dev/copy-vol`
  - There is no filesystem error on **os-vol**

**Comment by WebberHuang1118** [comment]
LH team seems to find the root cause https://github.com/longhorn/longhorn/issues/6899#issuecomment-1892012755,
We could re-verify it if related PR https://github.com/longhorn/sparse-tools/pull/96 could be merged in time.

**Comment by WebberHuang1118** [comment]
This should be fixed in LH v1.6.0, the root cause is analyzed in https://github.com/longhorn/longhorn/issues/6899#issuecomment-1892012755 by [ChanYiLin](https://github.com/ChanYiLin)

TL;DR, during reading out of backing image size but still within volume size, LH returns the wrong read size and this makes the return buffer cut incorrectly, as a result, it brings data inconsistency for volume export.

The related PR is https://github.com/longhorn/longhorn-engine/pull/986.

Additional information:
- The code path from volume export to replica read error is [VolumeExport()](https://github.com/longhorn/longhorn-engine/blob/f2418966602de5d3eb3ab7e7dfeb8dc38d214380/pkg/sync/rpc/server.go#L352-L382) -> [SyncContent()](https://github.com/longhorn/longhorn-engine/blob/f2418966602de5d3eb3ab7e7dfeb8dc38d214380/vendor/github.com/longhorn/sparse-tools/sparse/client.go#L129-L175) -> [syncContent()](https://github.com/longhorn/longhorn-engine/blob/f2418966602de5d3eb3ab7e7dfeb8dc38d214380/vendor/github.com/longhorn/sparse-tools/sparse/client.go#L177-L203) -> [processSegment()](https://github.com/longhorn/longhorn-engine/blob/f2418966602de5d3eb3ab7e7dfeb8dc38d214380/vendor/github.com/longhorn/sparse-tools/sparse/client.go#L270-L282) -> [syncDataInterval()](https://github.com/longhorn/longhorn-engine/blob/f2418966602de5d3eb3ab7e7dfeb8dc38d214380/vendor/github.com/longhorn/sparse-tools/sparse/client.go#L468-L494) -> [ReadDataInterval()](https://github.com/longhorn/longhorn-engine/blob/f2418966602de5d3eb3ab7e7dfeb8dc38d214380/vendor/github.com/longhorn/sparse-tools/sparse/file.go#L188-L200)

**Comment by lanfon72** [comment]
Verified this bug has been fixed.

Test Information
----
* Environment: **qemu/KVM 3 nodes**
* Harvester Version: **v1.4-f5fc452-head**
* **ui-source** Option: **Auto**

### Verify Steps
1. Install Harvester with any nodes
2. Create an image for VM creation
3. Create VM `vm1` and write data
4. Stop `vm1` and check its disk volume is detached completely
5. Export the disk volume into image `img1`
6. Add `img1` into `vm1` as additional data disk `disk-2`
7. Start `vm1`, VM should started successfully
8. Check data in additional disk `disk-2` is consistent with root disk
